policy: Add ICMP and ICMPv6 support for CNP and CCNP with a feature flag

chez-shanpu opened the pull request:
Related: #14609

Add `icmps` field to CNP and CCNP, and ICMP communication is allowed/disallowed by setting ICMP-type to that field.
Also this API supports ICMPv6 by setting "IPv6" to `family` field.

Internally, this API set ICMP policy by converting it into `port/protocol` format (`type/ICMP`).
Thus, this PR won't change the policy map layout.

Currentlly, this ICMP CNP feature is disabled by default because of the kernel complexity issue, and it is enabled when the "enable-icmp-rule" hidden flag is set.

---
Example yaml manifest:

```yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "cnp-icmp"
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
  - fromEndpoints:
    - matchLabels:
        org: empire
    icmps:
    - fields:
      - type: 8
      # ICMPv6
      - type: 128
        family: IPv6
```

chez-shanpu committed:
policy: Add ICMP and ICMPv6 API with feature flag

This commit implements ICMP/ICMPv6 API for CNP and CCNP.
The `icmp` field is added to CNP and CCNP manifests, and policy can be added by setting ICMP type to that field.
Currentlly, this ICMP CNP feature is disabled by default because of the kernel complexity issue, and it is enabled when the "enable-icmp-rule" hidden flag is set.

Example manifest:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "cnp-icmp"
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
  - fromEndpoints:
    - matchLabels:
        org: empire
    icmps:
    - fields:
      - type: 8
      # ICMPv6
      - type: 128
        family: IPv6
```

Signed-off-by: Tomoki Sugiura <cheztomo513@gmail.com>

christarazi commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,72 @@
+package api
```
Nit: We should add the copyright headers. Feel free to copy-paste from an existing file, but just be sure to update the year

chez-shanpu replied:
I see. Thank you :)

christarazi commented on pkg/policy/api/egress.go:
```diff
@@ -199,6 +209,16 @@ type EgressDenyRule struct {
 	ToPorts PortDenyRules `json:"toPorts,omitempty"`
+
+	// ICMP is a list of ICMP-type identified by type number
+	// which the endpoint subject to the rule is not allowed to connect to.
```
```suggestion
	// ICMP is a list of ICMP rules identified by type number
	// which the endpoint subject to the rule is not allowed to connect to.
```

Would that sound better?

christarazi commented on pkg/policy/api/egress_test.go:
```diff
@@ -308,6 +308,29 @@ func (s *PolicyAPITestSuite) TestIsLabelBasedEgress(c *C) {
+							{
+								Types: []ICMPType{
+									{
+										Type: 0,
```
It's probably better to assert against non-zero values, because it would give more confidence that the code has worked properly, as [Go will zero all values](https://tour.golang.org/basics/12) during linking. If the implementation is buggy, it's possible that the test could still pass because the values would be zeroed.

chez-shanpu replied:
OK I'll set the other number :)

christarazi commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+// ICMPType is a ICMP-type.
+type ICMPType struct {
```
Just curious why we've chosen to call this `ICMPType` rather than `ICMPRule`.

chez-shanpu replied:
What do you mean? I'm using `ICMPRule` in the line above.

christarazi replied:
Ah, nevermind. :)

christarazi commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,72 @@
+	// It should be 0-255 (8bit).
+	//
+	// +kubebuilder:validation:Maximum:=255
+	// +kubebuilder:validation:Minimum:=0
```
Nit: I think the extra colon near the end is not needed. I'm actually a bit surprised that the code gen tool didn't panic or something :)
```suggestion
	// +kubebuilder:validation:Maximum=255
	// +kubebuilder:validation:Minimum=0
```

chez-shanpu replied:
It's correctly work without the colon as you say, but It seems the correct format is with the extra colon.
ref: https://book.kubebuilder.io/reference/markers/crd-validation.html

christarazi replied:
Ah interesting. Probably worth going with what is already consistent in the codebase

chez-shanpu replied:
Yes, I'll remove them :)

christarazi reviewed (commented):
Looking good so far! Just a few minor comments.

chez-shanpu committed:
policy: fix `cilium bpf policy get` command output

When the protocol in policy map key is ICMP or ICMPv6 and port is 0, avoid protocol field getting changed to "ANY" in `cilium bpf policy get` command output

Signed-off-by: Tomoki Sugiura <cheztomo513@gmail.com>

chez-shanpu committed:
policy: Add ICMP and ICMPv6 supports to `cilium policy trace`

Add ICMP and ICMPv6 to api/v1 models and update policy trace.

Signed-off-by: Tomoki Sugiura <cheztomo513@gmail.com>

chez-shanpu committed:
policy: Add bpf ICMP policy support with the "ENABLE_ICMP_POLICY" flag

ICMP types are set to the destination port in the policy map's key struct, when "ENABLE_ICMP_POLICY" flag is enabled.

Signed-off-by: Tomoki Sugiura <cheztomo513@gmail.com>

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1488,6 +1488,31 @@ spec:
+                      items:
+                        description: ICMPRule is a list of ICMP-type.
+                        properties:
+                          types:
```
Is `types` the most appropriate name? If we latter extend this with ICMP codes, we will have something like:
```yaml
    - types:
       - type: 3
         code: 9
       - type: 3
         code: 10
```
Not sure of a better name :-/ Worst case we can raise the question on #development or #sig-policy.

chez-shanpu replied:
Yes, I don't think `types` is the best name.
However, currently we don't support ICMP codes or other header's field because there is no usecase, and also we don't have any plan to support them. Thus I think `types` is ok.

Or, we can name like `HeaderFields`or `attributes`?

What do you think?

christarazi replied:
`fields` is used in the [RFC](https://datatracker.ietf.org/doc/html/rfc792), so maybe that's a good option?

chez-shanpu replied:
I think it's good idea :)

pchaigno replied:
> we don't have any plan to support them.

We may support codes in the future.

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1488,6 +1488,31 @@ spec:
                             type: object
                         type: object
                       type: array
+                    icmp:
```
This should be `toICMPs` to match `toPorts` and other rules we already support.

chez-shanpu replied:
I wonder what the "to" means.
Since port number is different for destination and source, we need "to".
However ICMP type is not, so I don't think "to" is necessary.

pchaigno replied:
Oh, you're totally right! We probably still need the plural though, even if `icmps` is a bit ugly...

chez-shanpu replied:
How about `ICMPs`?

pchaigno replied:
I think the first letter needs to be lower case.

chez-shanpu replied:
OK, so let's change to `icmps`.

pchaigno replied:
We can always change latter (until the v1.11 release) based on feedback from others :-)

pchaigno commented on pkg/policy/api/rule_validation.go:
```diff
@@ -28,6 +28,7 @@ import (
 
 const (
 	maxPorts = 40
+	maxTypes = 40
```
```suggestion
 	maxICMPTypes = 40
```

pchaigno commented on pkg/policy/rule_test.go:
```diff
@@ -263,6 +264,108 @@ func (ds *PolicyTestSuite) TestL4Policy(c *C) {
 	c.Assert(egressState.selectedRules, Equals, 0)
 	c.Assert(egressState.matchedRules, Equals, 0)
+
+	// A rule for ICMP
```
I think it would be best to put the tests for ICMP in their own function. That will clarify what is failing in the test output.

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1563,6 +1563,32 @@ spec:
                             type: array
                         type: object
                       type: array
+                    icmp6:
```
I wonder if it would be worth wrapping `toICMPs` and `toICMP6s` under a single `toICMPs` field with a `family` subfield. That has the potential to significantly reduce code duplication. /cc @christarazi

christarazi replied:
Agreed, I think this is a good idea.

chez-shanpu replied:
@pchaigno use `family` field like this right?
```yaml
...
spec:
  ingress:
  - fromEndpoints:
    - matchLabels:
        org: empire
    toICMPs:
      family:
        toICMPs:
        - types:
          - type: 8
        toICMP6s:
        - types:
          - type: 128
```

pchaigno replied:
I was thinking of something like:
```yaml
spec:
  ingress:
  - fromEndpoints:
    - matchLabels:
        org: empire
    toICMPs:
    - fields:
      - family: IPv4
        type: 8
      - family: IPv6
        type: 128
```
Not sure what should be the value for `family:`. Maybe `ICMP` vs. `ICMPv6` is more appropriate?

Also, we could make `family: IPv4` the default for brevity.

chez-shanpu replied:
I think the unwrapped format is better.
As you said, we can reduce codes by wrapping them, but I like current separated style because it is simpler and easy to read.

pchaigno replied:
Then maybe we can find some other way to reduce code duplication?

chez-shanpu replied:
Ah... to reduce codes I think your idea is the best.
In my thought, ICMP and ICMPv6 are completely different things, so I wanted to separate. But I'm starting to think wrapped style is good.
Introduce `family` field makes the manifest complicated but not so much and the amount of codes we'll reduce is so much.

pchaigno replied:
And if it defaults to IPv4 family, then it shouldn't be more complicated for the most common cases.

I agree ideally it might be best to separate ICMP and ICMPv6 because they are different protocols but at the same time they both have types and codes :shrug: One could argue it's the same as having UDP and TCP under the same `toPorts` field.

If in doubt, we can always raise it on e.g. #sig-policy or #networkpolicy on Slack.

pchaigno commented on bpf/include/linux/in.h:
```diff
@@ -54,6 +54,8 @@ enum {
   IPPROTO_AH = 51,		/* Authentication Header protocol	*/
 #define IPPROTO_AH		IPPROTO_AH
+    IPPROTO_ICMP_V6 = 58,		/* Internet Control Message Protocol for IPv6	*/
+#define IPPROTO_ICMP_V6		IPPROTO_ICMP_V6
```
Nit: Seems to be some indentation issues here.

pchaigno replied:
There's still some misalignment here compared to the lines above, no?

chez-shanpu replied:
I fixed the left space, but the right space varies from line to line and I don't know where to fit it (so now I input same number of tabs).

pchaigno replied:
You should make it consistent with the lines above (i.e., the comment and definition should be aligned on previous lines).

chez-shanpu replied:
OK, I'll fix it :)

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,35 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+			return DROP_INVALID;
+
+		off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
```
We can load only the 1-byte ICMP type instead of loading the whole ICMP header from the packet.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,35 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+		off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
+			return DROP_INVALID;
+		key.dport = icmphdr.type << 8;
```
This shift needs a comment to explain what's happening here and why we're not shifting the value in userspace before inserting in the map.

pchaigno reviewed (commented):
Nice work! This is really neat :rocket: I only have minor comments inline, plus two other more general comments below:

**Regarding end-to-end tests which I triggered:** many are currently failing, but most look unrelated to your pull request. I think you just need to rebase against latest master. There is however one complexity issue with the 4.9 kernel; we can probably discuss that one once the first set of reviews is addressed.

**Regarding commit history:** I think the first 4 commits should be merged together. They are basically making the same-ish changes in the same places. Right now, reviewers need to review the same changes 4 times when they could simply do one pass over those files.
If you find another way to split those 4 commits without breaking the build, then it may be useful. Maybe CNP/CCNP changes can be introduced in one commit and the validation in another? Not sure if that would break anything.

christarazi commented on pkg/policy/rule.go:
```diff
@@ -665,7 +665,7 @@ func (r *rule) matches(securityIdentity *identity.Identity) bool {
 // ****************** EGRESS POLICY ******************
 
-func mergeEgress(policyCtx PolicyContext, ctx *SearchContext, toEndpoints api.EndpointSelectorSlice, toPorts api.PortsIterator, ruleLabels labels.LabelArray, resMap L4PolicyMap, fqdns api.FQDNSelectorSlice) (int, error) {
+func mergeEgress(policyCtx PolicyContext, ctx *SearchContext, toEndpoints api.EndpointSelectorSlice, toPorts, ICMP api.PortsIterator, ruleLabels labels.LabelArray, resMap L4PolicyMap, fqdns api.FQDNSelectorSlice) (int, error) {
```
Style nit: I was a bit confused seeing `ICMP` the variable being used as I thought it was somehow referring to `ICMP ICMPRules` from the rules struct. Let's keep this variable lower-case

christarazi reviewed (changes requested):
Nice work! 

I think that we can squash the first 4 commits into one because I think they all represent the same change. One thing that needs to be done is to increment the CRD schema version inside `pkg/k8s/apis/cilium.io/v2/register.go` because the addition of the `ICMP` piece to the polices.

I only have one minor comment besides Paul's comments.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,47 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+			return DROP_INVALID;
+
+		off = ((void *)ip6 - data) + ipv6_hdrlen(ctx, ETH_HLEN, &ip6->nexthdr);
+		if (ctx_load_bytes(ctx, off, &icmp6_type, sizeof(icmp6_type)) < 0)
```
I think we can merge the ICMP and ICMPv6 blocks because the `type` field is at the same offset for both. It may also help for the complexity issue on 4.9.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,47 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+		key.dport = (__u16)icmp_type;
+#else
+		key.dport = (__u16)(icmp_type << 8);
+#endif
```
I'd rather we make such checks in userspace, mostly because:
1. If we can move complexity to userspace without cost, we should.
2. We will likely want to add support for ICMP codes later on and in that case the type needs to be on the 8 upper bits.

chez-shanpu replied:
I agree with you :)
However I don't know how to check it in the userspace.
Because we get `icmp_type` from `ctx_load_bytes()` and it's always 8bit, so we have to expand it to 16bits in bpf.
Do you have any idea?

pchaigno replied:
Actually, I don't think we support big-endian architectures in the agent at all: https://github.com/cilium/cilium/blob/v1.10.1/pkg/byteorder/byteorder.go#L70-L79. So it's probably fine to only support little-endian here.

chez-shanpu replied:
Sure, I'll remove big-endian case.

joestringer replied:
Just a heads up, the statement about not supporting big-endian is incorrect. The endian check is done inside the swap functions:

https://github.com/cilium/cilium/blob/e6f34c3c98fe2e247fde581746e552d8cb18c33c/pkg/byteorder/byteorder.go#L42-L43

pchaigno reviewed (commented):
Two minor comments below. The first may allow us to get rid of the complexity issue on 4.9 :crossed_fingers: 

Don't forget to update the example policies in your first commit and in this pull request's original message.

christarazi commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,108 @@
+	//
+	// +kubebuilder:default=`IPv4`
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern=`IPv(4|6)`
```
While checking it against the pattern will work, we can use the `Enum` validation because we only expect two values:

```suggestion
	// +kubebuilder:validation:Enum=IPv4;IPv6
```

christarazi commented on pkg/policy/api/ingress.go:
```diff
@@ -116,6 +116,17 @@ type IngressRule struct {
+	// type 8 ICMP connections.
+	//
+	// +kubebuilder:validation:Optional
+	ICMP ICMPRules `json:"icmps,omitempty"`
```
If the JSON name is going to be plural, then let's make the field name also plural.
```suggestion
	ICMPs ICMPRules `json:"icmps,omitempty"`
```

christarazi reviewed (commented):
Looking better!

I don't see the increment of the CRD schema version inside `pkg/k8s/apis/cilium.io/v2/register.go`.

chez-shanpu commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -34,7 +34,7 @@ const (
 	// Maintainers: Run ./Documentation/check-crd-compat-table.sh for each release
 	// Developers: Bump patch for each change in the CRD schema.
-	CustomResourceDefinitionSchemaVersion = "1.23.3"
+	CustomResourceDefinitionSchemaVersion = "1.24.0"
```
@christarazi I already updated schema version, but it's not enough...?

christarazi replied:
Ah, I somehow didn't notice it. Let's update it to 1.23.4 instead.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,37 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+		__u8 icmp_type;
+		__u32 off;
+
+		if (proto==IPPROTO_ICMP){
```
Watch out for spaces here and in the `}else{` :-)

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -123,6 +123,37 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+			if (!revalidate_data(ctx, &data, &data_end, &ip6))
+				return DROP_INVALID;
+			off = ((void *)ip6 - data) + ipv6_hdrlen(ctx, ETH_HLEN, &ip6->nexthdr);
+		}
```
Doesn't it work if you just use `struct iphdr` for both ICMP and ICMPv6 (with a comment to explain they have the type at the same offset)?

chez-shanpu replied:
I tried, but it didn't work.
I think the length of ip4 and ip6 is different, so the value of `off` is different.

pchaigno replied:
Ah, right. You still need the two structures for `ip6->nexthdr` :disappointed: 
You could still move the `revalidate_data` out but I'm not sure that will buy us a lot w.r.t. the complexity. The compiler should be able to do the same.

pchaigno reviewed (changes requested):
The unit tests are still failing (cf. https://app.travis-ci.com/github/cilium/cilium/jobs/521407004).

christarazi commented on bpf/lib/conntrack.h:
```diff
@@ -615,18 +616,19 @@ static __always_inline int ct_lookup4(const void *map,
 			tuple->dport = 0;
 
+			/* convert from unsigned char to unsigned short considering byteorder(little-endian) */
+			tuple->sport = (__u16)(type << 8);
```
Why is casting not enough to convert from 1 byte to 2 bytes? What's special about the other 8 bits? That would be good to clarify in this comment.

chez-shanpu replied:
This is because of byte order. In the little-endian case, for example 2byte data "AB" convert to "BA", so we should shifting the "type" not just casting. 
I commented about it in the line above, but it may not be the best explanation. So, you mean it should be more specific?

christarazi commented on bpf/lib/policy.h:
```diff
@@ -100,28 +100,43 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+				return DROP_INVALID;
+		}
+
+		key.dport = (__u16)(icmp_type << 8);
```
Could you add the clarifying comment here about the shifting and the byteorder?

pchaigno commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,108 @@
+	return len(ir)
+}
+
+// GetPortProtocols generate PortProtocol slice from ICMPRule and return it.
```
Nit: 
```suggestion
// GetPortProtocols generates PortProtocol slice from ICMPRule and return it.
```
Same for the other functions below.

pchaigno commented on test/k8sT/Policies.go:
```diff
@@ -391,6 +399,144 @@ var _ = SkipDescribeIf(func() bool {
 			res.ExpectFail("Unexpected connection from %q to 'http://%s/private'",
 				appPods[helpers.App3], clusterIP)
+
+			By("Testing ICMP rules")
```
Note I don't think the policies loaded by previous tests were all removed. Maybe we should fix that in a separate commit? Otherwise the ICMPs could be allowed by some L3 rule.

chez-shanpu replied:
I agree, but what does "in a separate commit" mean? Don't we fix and update the existing commit?
Also, do you think we should fix it by using another `It()` for ICMP testing?

pchaigno replied:
> I agree, but what does "in a separate commit" mean?

IMO it's not really an issue introduced by your code, so it could be in a separate commit.

>  Also, do you think we should fix it by using another `It()` for ICMP testing?

Another `It` block would also work. It has the advantage of clearly indicating what failed when it failed :-)

chez-shanpu replied:
I see, so now I'll divide it by another `It`.
And if an issue occur, let's fix it in a separate commit : )

christarazi replied:
Removing the policies at the end of the test is already happening from the 
```go
		AfterEach(func() {
			cmd := fmt.Sprintf("%s delete --all cnp,ccnp,netpol -n %s", helpers.KubectlCmd, namespaceForTest)
			_ = kubectl.Exec(cmd)
		})
```
so I don't think an additional commit is necessary.

chez-shanpu replied:
Neither do I. Now I divided ICMP tests and others by `It` blocks, so probably we don't have to worried about that since the `AfterEach` was executed before ICMP `It` block starts.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -100,28 +100,43 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
 
-		if (!revalidate_data(ctx, &data, &data_end, &ip4))
-			return DROP_INVALID;
+		if (proto==IPPROTO_ICMP) {
```
```suggestion
		if (proto == IPPROTO_ICMP) {
```

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -100,28 +100,43 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+				return CTX_ACT_OK;
 #endif /* ALLOW_ICMP_FRAG_NEEDED */
+			icmp_type = icmphdr.type;
+		} else{
```
```suggestion
		} else {
```

pchaigno reviewed (approved):
Only nits from me. Nice work on the bit of C code refactoring!

christarazi commented on bpf/lib/policy.h:
```diff
@@ -100,28 +100,43 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
+			 * of ICMP type 3 code 4 - Fragmentation Needed.
+			 */
+			if (icmphdr.type == ICMP_DEST_UNREACH &&
+				icmphdr.code == ICMP_FRAG_NEEDED)
```
```suggestion
			if (icmphdr.type == ICMP_DEST_UNREACH &&
			    icmphdr.code == ICMP_FRAG_NEEDED)
```

christarazi reviewed (approved):
LGTM as well. Nice work!

qmonnet commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,108 @@
+	return len(ir)
+}
+
+// GetPortProtocols generates PortProtocol slice from ICMPRule and return it.
```
`returns`

qmonnet commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,108 @@
+	return pps
+}
+
+// GetPortRule generates PortRule from ICMPRule and return it.
```
`returns`

qmonnet commented on pkg/policy/api/icmp.go:
```diff
@@ -0,0 +1,108 @@
+func (ir ICMPRules) Iterate(f func(pr Ports) error) error {
+	for i := range ir {
+		err := f(&ir[i])
+		if err != nil {
```
`if err := f(&ir[i]); err != nil {`?

qmonnet commented on Documentation/policy/language.rst:
```diff
@@ -678,6 +678,56 @@ only be able to emit packets using TCP on port 80, to any layer 3 destination:
         .. literalinclude:: ../../examples/policies/l4/l4.json
 
+Limit ICMP/ICMPv6 types
+--------------------------
```
```suggestion
Limit ICMP/ICMPv6 types
-----------------------
```

Also, note that there are titles at different levels under the `Layer 4 Examples`. The current hierarchy is:
```
Layer 4 Examples
        - Limit ingress/egress ports
                - Example (L4)
                - Labels-dependent Layer 4 rule
                - CIDR-dependent Layer 4 Rule
```

With your changes, you turn it like this:
```
Layer 4 Examples
        - Limit ingress/egress ports
                - Example (L4)
        - Limit ICMP/ICMPv6 types
                - Example (ICMP/ICMPv6)
                - Labels-dependent Layer 4 rule
                - CIDR-dependent Layer 4 Rule
```

This places the last two sub-subsections under your ICMP documentation, which is likely not what you want. You should probably move your subsection at the end, below the `CIDR-dependent Layer 4 Rule`.

chez-shanpu replied:
Oh, thank you!
Yes, I'll put it under the CIDR section :)

qmonnet commented on Documentation/policy/language.rst:
```diff
@@ -678,6 +678,56 @@ only be able to emit packets using TCP on port 80, to any layer 3 destination:
+~~~~~~~~~~~~~~~~~~~~~
+
+The following rule limits all endpoints with the label ``app=myService`` to
+only be able to emit packets using ICMP with type 8 and ICMPv6 with type 128, to any layer 3 destination:
```
Please wrap the line.

qmonnet reviewed (changes requested):
Looks great overall, just a few nits below.

qmonnet reviewed (approved):
Thanks for the changes, looks good to me! :)

twpayne commented on pkg/policy/api/rule_validation.go:
```diff
@@ -401,6 +414,20 @@ func (pp *PortProtocol) sanitize() (isZero bool, err error) {
 	return isZero, err
 }
 
+func (ir *ICMPRule) sanitize() error {
```
nit: `sanitize` implies that `ir` is sanitized, i.e. modified. `verify` would probably be a better name.

twpayne commented on pkg/policy/rule.go:
```diff
@@ -421,6 +421,33 @@ func mergeIngress(policyCtx PolicyContext, ctx *SearchContext, fromEndpoints api
 		return nil
 	})
+	if err != nil {
+		return 0, err
```
Should this be `return found, err` for consistency with the previous code in this function?

twpayne commented on pkg/policy/rule.go:
```diff
@@ -719,6 +746,29 @@ func mergeEgress(policyCtx PolicyContext, ctx *SearchContext, toEndpoints api.En
 	},
 	)
+	if err != nil {
+		return 0, err
```
Same comment re `return found, err`.

pchaigno commented on Documentation/policy/language.rst:
```diff
@@ -727,6 +727,57 @@ ports other than port 80.
+It restricts the ability of an endpoint to emit and/or receive packets on a
+particular ICMP/ICMPv6 type (currently ICMP/ICMPv6 code is not supported).
+If any ICMP policy is specified, layer 4 and ICMP communication will be blocked
+unless it's related to a connection that is otherwise allowed by the policy.
```
I think we should remove this sentence. It's just the usual "loading a policy switches policy enforcement into default-deny". It's also already explained in the previous section where it makes more sense to insist on that (given some folks expect ICMP to always be allowed).

chez-shanpu replied:
Yes, but considering the people who only read this ICMP section, we should leave this sentence?

pchaigno reviewed (commented):
The documentation looks good. I only have one nit below on that.

I've done some additional testing and discovered that the below policy allows both TCP/443 and ICMP/8. I think that may be a bit confusing to users, and we probably should just reject such policies instead.
```yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "cnp-icmp"
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
  - fromEndpoints:
    - matchLabels:
        org: empire
    icmps:
    - fields:
      - type: 8
    toPorts:
    - ports:
      - port: "443"
        protocol: TCP
```

christarazi commented on pkg/policy/api/rule_validation.go:
```diff
@@ -230,6 +241,10 @@ func (e *EgressRule) sanitize() error {
 	}
 
+	if len(e.ICMPs) > 0 && len(e.ToPorts) > 0 {
+		return fmt.Errorf("The ICMPs block must only be present without any other rules. Define a separate rule to allow other kinds of traffic.")
```
Ah, I didn't realize that only `ToPorts` would be relevant here. In that case, we could probably make this msg a bit more specific.

```
The ICMPs block may only be present without ToPorts. Define a separate rule to use ToPorts.
```

christarazi commented on pkg/policy/api/rule_validation_test.go:
```diff
@@ -686,6 +686,102 @@ func (s *PolicyAPITestSuite) TestTooManyPortsRule(c *C) {
+	}
+
+	err := ingressICMPWithPort.Sanitize()
+	c.Assert(err, NotNil)
```
I think we can use the `ErrorMatches` matcher here to assert on the actual error returned

christarazi commented on pkg/policy/api/rule_validation.go:
```diff
@@ -219,6 +235,14 @@ func (e *EgressRule) sanitize() error {
 
+	if len(e.ICMPs) > 0 && !(viper.GetBool(option.EnableICMPRule)) {
+		return fmt.Errorf("ICMP rules can only be applied when the \"%s\" flag is set", option.EnableICMPRule)
+	}
```
```suggestion
	if len(e.ICMPs) > 0 && !option.Config.EnableICMPRule {
		return fmt.Errorf("ICMP rules can only be applied when the %q flag is set", option.EnableICMPRule)
	}
```

christarazi replied:
We don't need to import viper to do this check, all the options are loaded in `option.Config`.

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -982,6 +982,10 @@ func init() {
 
+	flags.Bool(option.EnableICMPRule, false, "Enables ICMP/ICMPv6 cilium network policy rules")
+	flags.MarkHidden(option.EnableICMPRule)
+	option.BindEnv(option.EnableICMPRule)
```
```suggestion
	flags.Bool(option.EnableICMPRules, false, "Enable ICMP-based rule support for Cilium Network Policies")
	flags.MarkHidden(option.EnableICMPRules)
	option.BindEnv(option.EnableICMPRules)
```

christarazi commented on pkg/option/config.go:
```diff
@@ -954,6 +954,10 @@ const (
+
+	// EnableICMPRule enables ICMP/ICMPv6 ciilum network policy rules
+	// todo: remove this after the kernel complexity issue is resolved
+	EnableICMPRule = "enable-icmp-rule"
```
```suggestion
	EnableICMPRules = "enable-icmp-rules"
```

christarazi reviewed (changes requested):
A few minor changes are required regarding the feature flag.

christarazi commented on pkg/defaults/defaults.go:
```diff
@@ -430,4 +430,8 @@ const (
 	// Defaults to false to retain prior behaviour of not routing external packets to ClusterIPs.
 	ExternalClusterIP = false
+
+	// EnableICMPRules enables ICMP/ICMPv6 ciilum network policy rules
```
```suggestion
	// EnableICMPRules enables ICMP-based rule support for Cilium Network Policies.
```

christarazi commented on pkg/defaults/defaults.go:
```diff
@@ -430,4 +430,8 @@ const (
 	ExternalClusterIP = false
+
+	// EnableICMPRules enables ICMP/ICMPv6 ciilum network policy rules
+	// todo: remove this after the kernel complexity issue is resolved
```
I think we can remove this notes from the PR. Instead, after the PR is merged, let's open a new issue and reference this PR in it.
